### DISCUSSION TITLE
Is there a way to prevent it from being converted to .classdata?

### DISCUSSION THREAD

**Discussion by zerocooldog** [open-telemetry/opentelemetry-java-instrumentation#12813] [Q&A]
There are cases where I need to access classes converted to .classdata from an external library.

There is a way to register the class that requires external access in the bootstrap package, but this method cannot be used in this case.

Is there an alternative way to access them in such a case?

**Comment by steverao** [comment]
You can refer to [related design introduction](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/javaagent-structure.md#agent-class-loader) about `.classdata`. I am not sure about your specific demand. If you want to access them outside the Agent class loader, you need to use Agent class Loader of OTel Java Agent. You can refer to [related using case](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/7b69e56cb632c4dd08c99ab3d3ba01a603cd236d/javaagent-bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/AgentInitializer.java#L192-L200) in OTel Java Agent.

**Reply by zerocooldog** [reply]
@steverao I was asking the question because we were implementing a new library required for our work, and I couldn't import the .classdata file of the OTel Java Agent into the IDE. I personally solved the problem by implementing a separate Proxy class. Thank you for your answer.

**Reply by laurit** [reply]
> and I couldn't import the .classdata file of the OTel Java Agent into the IDE

instead of using the agent jar as a dependency use the module that contains the class, such as https://central.sonatype.com/artifact/io.opentelemetry.javaagent/opentelemetry-javaagent-tooling

**Reply by zerocooldog** [reply]
@laurit 

I'm not having any issues importing the Java classes provided by otel. My question is about importing a class that I've implemented myself.

I'm implementing an agent using the distro approach for the Otel Java agent, and I've written the necessary logic in a separate instrumentation module.

However, the class I've written is also being transformed into a .classdata file, preventing my IDE from importing it. 

This is why I asked if there's a way to skip the .classdata transformation.

![image](https://github.com/user-attachments/assets/bb9ad997-b66a-49ce-8d90-27f69e35f41f)

**Reply by laurit** [reply]
> However, the class I've written is also being transformed into a .classdata file, preventing my IDE from importing it.

`.class` -> `.classdata` rename happens during assembling the agent jar. You should have the original jars that contain the classes as the were before shading and renaming.

**Reply by zerocooldog** [reply]
@laurit I'm well aware of that. However, I asked because when developing an external library, the already built agent.jar must be added as a dependency. I solved this problem in another way.

Thanks for the answer.

**Reply by laurit** [reply]
> the already built agent.jar must be added as a dependency

Why would you do that? As you have discovered yourself it doesn't work, use the libraries that the agent is assembled from instead.

**Reply by zerocooldog** [reply]
@laurit 

The reasons I used it that way are as follows:

1. I worked on it because I needed custom metric logic in the otel Java agent, but I wasn't aware of the .classdata conversion process. Later, I realized it didn't work when developing in the IDE and importing it.

2. Even if I add the agent project as a dependency, I cannot use the shaded functionality in the actual project I'm working on, so I can't match the packages exactly.

Finally, I could have solved it by writing custom metric logic and using typeInstrumentations to inject bytecode, but in my case, since I needed to capture metric values at various points in the middle of the logic, rather than just at the method start and end, I had no choice but to apply this approach.

In conclusion, I added the class to the bootstrap package and worked on it.

**Reply by laurit** [reply]
> Even if I add the agent project as a dependency, I cannot use the shaded functionality in the actual project I'm working on, so I can't match the packages exactly.

You could use the same approach as the sample distro. Depend on the unshaded agent components and apply the same shading rules as the agent.

**Reply by zerocooldog** [reply]
@laurit I thought that I could probably reference a the distro/custom module, but since the project I'm currently working on is not heavily dependent on the agent, I ruled out that approach.

Anyway, thanks for your long responses. I'll wrap it up here. :)
